### ISSUE TITLE
UX of virtual branch commit states (discussion)

### ISSUE THREAD

**Issue by krlvi** [gitbutlerapp/gitbutler#3121]
## Background:

The nature of branches having a local and remote equivalent means that commits can be in multiple states.
1. If the local and remote are identical, there’s nothing to do
2. If the local has commit(s) that are not on the remote, those commits can be pushed
3. If the remote has commit(s) (from another user or the GitHub GUI) that are not local, they can be pulled
4. If the local branch has non-punished commits and the remote has non-pulled commits the branches are diverged

I find this graphic by Julia Evans to be quite nice at illustrating this:

![diverged](https://github.com/gitbutlerapp/gitbutler/assets/4030927/c8ebd9b0-4b98-4ad4-9f1e-de9e3507d359)

## Problem

The lanes GitButler renders for virtual branches don’t represent the `need to pull` and `diverged` states clearlly. To illustrate this, I made a graphic showcasing the way the 4 states are shown by GitButler

## Current state in GitButler

Below are 4 screenshots next to each other, covering the 4 states mentioned previously and how they are represented within GitButler virtual branches, alongside some commentary. I generated these states by making a commits using the GitHub website. 
 
<img width="1570" alt="current_lanes" src="https://github.com/gitbutlerapp/gitbutler/assets/4030927/b25b4d7b-3e84-408f-a909-0b66f0c61273">

Comments from the graphic 👆 (the numbers in blue)
- (1) “Upstream commits” duplicate what is local
- (2) “Merge” is misleading as this would mean simply pulling the remote commit
- (3) “Remote branch” is a misleading term as here we show only local commits
- (3) “Force push to remote” doesn’t make sense as this is in the context of remote commits
- (3) “Force push” may have unintended consequences (losing the remote commit)
- (4) Duplicated from below
- (5) “Merge” is a bit misleading
- (6) “Force push to remote”, unclear difference from the other force push button, possibly unintended consequences 
- (7) Possibly unintended consequences 

## Lets discuss 💬

The purpose of this GH issue is to invite a discussion and maybe go over some drafts / ideas that come up. @schacon, @mtsgrd, @Qix- @PavelLaptev let me know if I have missed something

**Comment by mtsgrd** [comment]
@krlvi I should perhaps know the answer by now, but what is the difference between "upstream commits" and commits on "remote branch"?

**Comment by krlvi** [comment]
> @krlvi I should perhaps know the answer by now, but what is the difference between "upstream commits" and commits on "remote branch"?

I view the current state as a UI (UX) bug. In terms of what goes in each group: 
- `remote branch` appear to be showing commits that are both local **and** remote (with the teal and purple color) - this is misleading
- `upstream commits` appear to be showing the branches that on the remote.

(very likely the way we have the distinction now is something for us to fully redo)

**Comment by schacon** [comment]
Trying to illustrate, let's say you start a new branch and commit twice (A and B), you have two "local" commits:

![CleanShot 2024-03-12 at 11 47 35@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/3cec279a-21ec-4531-83ac-fa176a51ab45)

Then you push the branch to GitHub, now those commits are on the server as well, so we let you know that by labeling them as being on the "remote branch":

![CleanShot 2024-03-12 at 11 49 22@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/093d94e7-23a3-425a-ac15-0723d29a7a7e)

Ok, now you commit on your branch again but haven't pushed it, so now your commits are in two different groups. One that is local only and one that is pushed:

![CleanShot 2024-03-12 at 11 50 11@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/dca8b014-02f6-4f88-a286-834576ba54b4)

Finally, someone _else_ pushes work to your GitHub branch (something not created locally by you). Now you have an "upstream" commit.

![CleanShot 2024-03-12 at 11 50 46@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/3bd7d674-3804-4a43-9432-5bc3bbe724ce)

Now lets say you merge in the upstream commit:

![CleanShot 2024-03-12 at 11 55 38@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/3451a32f-bc64-42c9-a7b4-2ffddf89232b)

Now you have a few local commits (the merged one included). Now you push, everything is "remote branch", because all the commits are on the server:

![CleanShot 2024-03-12 at 13 06 33@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/b185df46-75d1-4baa-b5b1-e2e47dc2827a)

**Comment by schacon** [comment]
I think this is the minimally important distinction. Is a commit:

1. present locally
2. present on a server

If it's *1* only, it's in the "local" group

If it's *1* _and_ *2*, it's "remote branch" group

If it's _only_ *2*, it's "upstream" group

We can rename them or differently visualize them, but there are practical implications to each of these groups. 

* "local" grouped commits can be rewritten without needing to force push
* "remote branch" grouped commits are possibly on someone else's computer and more care needs to be taken
* "upstream" commits are not present in your working copy and could be conflicting and will _need_ to be integrated somehow

**Comment by krlvi** [comment]
Thanks for that illustration! I was thinking - we currently linearize the sequence, the way we show it, even when there is a merge. Perhaps that's a big source of confusion. 

Since we allow merges (and many people prefer it) perhaps we can try visualising it in a non-linear way. Even as a rebaser-person I'd find it perhaps useful.

**Comment by schacon** [comment]
It could be interesting to pull the graph component out to the side and illustrate the topology a little bit more, rather than flattening and simplifying the relationships (esp where we have merges). `jj` does this in an nice way:

![CleanShot 2024-03-12 at 14 01 37@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/fd27ee73-6920-441d-a1bf-5ecb60c963a3)

It would basically be `git log --graph --oneline [vbranch]` ish output, but prettier and a little more clear.

**Comment by schacon** [comment]
@KroniK907 in that case, the old commits become state 3 ("upstream") because they are on the server under that branch but not local. the new local work becomes state 1 ("local") as they are not upstream. You _could_ merge them back in, the way they're done in the diagram, but more likely what you want to do is force push. 

You're correct though, when that becomes problematic is when someone _else_ pushes on top, so you want to effectively cherry pick from upstream and then force push.

**Comment by krlvi** [comment]
> I think this is the minimally important distinction. Is a commit:
> 
> present locally
> present on a server

I took the example screenshots from the parent post and cropped out just the commits, leaving off the groups they were in. I also removed the duplications and just spelled out in english where the commit is. 

It's crazy how much simpler the same example becomes when we skip the groups and just spell out what is where, and no duplicates:

<img width="1570" alt="commits" src="https://github.com/gitbutlerapp/gitbutler/assets/4030927/409b5413-24c8-4018-9242-a2588a8e7bb3">

Maybe we can design something that communicates the thing that in english using shapes and colors and making it obvious at a glance what is where?

### Graph idea

We could in theory draw a mini graph of the commits like this:
<img width="385" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/4030927/03a8fdca-f038-48ec-81d4-922789be7174">
(the order of `commit 3` and `commit from remote 2` I presume comes from the timestamp, and of course it would be important to be able to clearly distinguish at a glance what is on my computer and what is not)

**Comment by krlvi** [comment]
> It would basically be `git log --graph --oneline [vbranch]` ish output, but prettier and a little more clear.

I tried the following while my vbranch is in the 4th state (diverged)
`git log --graph --oneline main..$(git show-ref refs/gitbutler/testing)`

```
* dafd306 Commit 3
* c0a106e Commit from remote
* 250db6c Commit 2
* fdd15ea Commit 1
* 828305c (origin/main) Merge pull request #12 from krlvi/Virtual-branch
```
If i log the origin one `git log --graph --oneline main..origin/testing`:

```
* 884e7d0 (origin/testing) Commit from remote 2
* c0a106e Commit from remote
* 250db6c Commit 2
* fdd15ea Commit 1
* 828305c (origin/main) Merge pull request #12 from krlvi/Virtual-branch
```

I guess it is not showing them as separate lanes because there they are not in the same place (one is in remotes).
@schacon is there a flag to have them show as 2 lanes next to each other?

**Comment by kiprasmel** [comment]
@krlvi 

> If i log the origin one `git log --graph --oneline main..origin/testing`:
> I guess it is not showing them as separate lanes because there they are not in the same place (one is in remotes).
> is there a flag to have them show as 2 lanes next to each other?

your syntax `main..origin/testing` is saying "everything in `origin/testing`, excluding everything in `main` (i.e. everything in origin/testing, up until merge-base with `main`).

what you want is: to see unique commits from both (symmetric difference). the syntax is `main...origin/testing` (note the 3 dots instead of 2).

see "SPECIFYING RANGES" in `man gitrevisions`

**Comment by krlvi** [comment]
okay, just got some help from Scott - this one gave what i wanted `git log --graph --oneline $(git show-ref refs/gitbutler/testing) origin/testing ^origin/main`:

```
* 884e7d0 (origin/testing) Commit from remote 2
| * dafd306 (testing) Commit 3
|/
* c0a106e Commit from remote
* 250db6c Commit 2
* fdd15ea Commit 1
```

**Comment by krlvi** [comment]
@PavelLaptev FYI, this graph will never have more than two tips (the diverged case, state 4)

**Comment by PavelLaptev** [comment]
@krlvi got it. Thank you for the explanation. I'll prepare some drafts :-)

**Comment by PavelLaptev** [comment]
Made a draft. Thoughts?

<img width="1515" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/156cb9f3-6884-4241-b7d4-79fc9fec95e9">

**Comment by PavelLaptev** [comment]
Drafts in Figma https://www.figma.com/file/FbeLt0yjY9RiNn8MXUXsYs/Client-Design?type=design&node-id=2100%3A42661&mode=design&t=lerXeTmJGwqsI4gg-1

![image](https://github.com/gitbutlerapp/gitbutler/assets/18498712/c8e475da-4b40-4a7c-a086-658564d8d462)

We can show additional info on hover for the commit card and the avatar:

<img width="724" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/bc3d060b-7a61-4e75-ae50-d543a13d420c">
<img width="691" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/e566f2f9-02d7-408c-a3d9-ecd387224979">

**Comment by krlvi** [comment]
i really like these drafts! as we discussed irl - we could explore showing in a subtle way one more "node" in the graph for the base of this branch. This is less important than the other things so it can be more subtle, but it could be nice to be able to hover on it and see what was the base here.

<img width="523" alt="Screenshot 2024-03-18 at 15 34 21" src="https://github.com/gitbutlerapp/gitbutler/assets/4030927/c6b14d37-3190-4be7-8ca9-f601410a1f8f">

But now that I am thinking about it, this would be duplicated for all vbranches in the workspace.... so perhaps this does not belong in the lane, maybe it needs to be somehow outside?

**Comment by schacon** [comment]
I really like this format, but there is something that we do now that is not reflected here that could be a bit confusing, which is indicate that the upstream-only commits are not incorporated into our working directory. In our current UI, upstream-only commits are _above_ the uncommitted changes and everything those changes depend on (ie, are merged locally) are below that block.

We should indicate this somehow. here, you are interspersing the commits by time of commit, but I don't think that's nearly as important as wether it's merged into our working directory or it's remote only (ie, pending).

Perhaps we force all the upstream work to the top and have a dividing line, or even move them again above the uncommitted changes section. I'm not sure what works better, but they should be a removed block that is clearly unincorporated.

**Comment by krlvi** [comment]
@schacon I was thinking that the red/orange color, together with the the absence of a merge of the two tracks indicates upstream only. Do you mean always keeping the upstream bit on the top to further highlight that like this mockup 👇 ?

<img width="710" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/4030927/2b7ed28d-8a62-48d2-bdc8-688e08dd7567">

**Comment by schacon** [comment]
Exactly. I think since there's three colors, intermixing them would be confusing as to what is in your working directory and what is not. The upstream ones still need to be integrated and that should be clear. Also, we'll need a UI for that (merge/rebase these), which we're missing here. Those upstream commits are a really different animal and this interface makes them seem equivalent.

**Comment by krlvi** [comment]
> Also, we'll need a UI for that (merge/rebase these), which we're missing here. 

wouldn't the button below cover those cases? it changes it's mode (merge upstream / [force]push) with an icon matching the color + highlighting the commits to be merged / pushed on hover

**Comment by schacon** [comment]
Oh sorry, I missed that. Yeah, that could work.

Though, can we do a mockup where we have "merge upstream" directly under the upstream group and "push to remote" under the local group?

**Comment by krlvi** [comment]
I would also be interested in seeing that. But I wonder if we could somehow do it with buttons that are smaller than those big full-width ones, somehow it feels that they break the continuity / flow for me when they are in between other stuff (like in the current implementation)

**Comment by schacon** [comment]
Something like this (sorry for the horrible photoshop):

![CleanShot 2024-03-19 at 16 28 29@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/db9213ce-aa86-42e5-b079-62a0fe5e84fc)

We should always group these together, and it seems a bit clearer that there is an action that is performed on that specific group. Also, there are then fewer dropdown options and we can keep the last one easily. Like, the top would be either "Merge" or "Rebase" and the user would probably always want that option. The locals would either be "push" or "open PR" (if no PR yet) and again would always want to default to that.

**Comment by schacon** [comment]
btw, I really like this where it is a single line, but it bends for the local ones:

![CleanShot 2024-03-19 at 16 31 00@2x](https://github.com/gitbutlerapp/gitbutler/assets/70/622b5012-632b-479b-acfe-b5a8abfcb5d6)

It's a nice way to visually distinguish these. Also, we should remember that _most_ of the time, this is all that will be there, there will much more rarely be upstream work.

**Comment by PavelLaptev** [comment]
Groups are cool, but instead of performing multiple actions on each group, let's focus on the most suitable action for the situation.

For instance, if there are upstream commits, we wouldn't recommend a "force push" (you also can't "push to remote" in this case) as the immediate action; instead, the preferable action would be to merge first. Let's explore using a dropdown menu for this purpose, similar to the concept in GitHub Desktop. It helps by suggesting the most common actions, making workflows smoother.

Let me make some new drafts 🙂 I'll try to make something today 

<img width="470" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/8d3cd67a-5e89-4505-8821-948c8289b1c0">

**Comment by krlvi** [comment]
There is one interesting state i just remembered of which would be cool if we can represent nicely - the case when a developer chooses to rebase the branch and they have previously pushed. In this case, the commit content is the same but the hashes are different due to the different base locally and on the remote.

It could be cool if we can communicate that without showing it as a divergence. Or maybe divergence is the only correct way of showing it?

**Comment by schacon** [comment]
Divergence in this case is the only correct way to show it. What might be valuable is to show a partial sha in these cases (or maybe even generally?)

**Comment by krlvi** [comment]
> Divergence in this case is the only correct way to show it. What might be valuable is to show a partial sha in these cases (or maybe even generally?)

In this case, if i have 10 commits that I have already pushed, and then choose to rebase my working directory, before I (force) push, I will have 20 commits on screen, with the upstream ones shown on top, right? I wonder if we can improve on that situation.

Btw, what is a partial sha?

**Comment by schacon** [comment]
maybe if we can match on identical message (or we could be spiffy and extract a patch-id from each, but there are corner cases that fuck both of those approaches - commit messages are generally fine and fast), we could do some viz where we collapse the rebased ones and mostly show the new ones. I think this is rather common, esp the more we add rebasing methods.

**Comment by schacon** [comment]
> Btw, what is a partial sha?

Just showing the first 6-8 chars, rather than all 40 chars.

**Comment by krlvi** [comment]
Here's a crazy idea - what if we use the base as an anchor for this. When a user rebases, we show the commits only once but with nodes in both tracks (remote and local) (illustration 3.). 
We can visualise that there is divergency by:
- on hover on the node we show the shas
- the two tracks have different bases - the one that is local we can hover over, the remote branch has a line that extends further down (the base is somewhere off screen)

![IMG_4231](https://github.com/gitbutlerapp/gitbutler/assets/4030927/baf924e7-9fe2-4cb4-ab42-adaabde04a4c)

What we could gain from this is simpler / more compact representation, and also communicate that it's not that the commit content is different but it's the base that actually differs.

WDYT?

**Comment by kiprasmel** [comment]
@krlvi isn't this exactly what `git range-diff` does?

see also https://github.com/kiprasmel/git-rebase-diff

**Comment by PavelLaptev** [comment]
> Here's a crazy idea - what if we use the base as an anchor for this. When a user rebases, we show the commits only once but with nodes in both tracks (remote and local) (illustration 3.). We can visualise that there is divergency by:
> 
> * on hover on the node we show the shas
> * the two tracks have different bases - the one that is local we can hover over, the remote branch has a line that extends further down (the base is somewhere off screen)
> 
> ![IMG_4231](https://private-user-images.githubusercontent.com/4030927/314194614-baf924e7-9fe2-4cb4-ab42-adaabde04a4c.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMTY1ODcsIm5iZiI6MTcxMTMxNjI4NywicGF0aCI6Ii80MDMwOTI3LzMxNDE5NDYxNC1iYWY5MjRlNy05ZmUyLTRjYjQtYWI0Mi1hZGFhYmRlMDRhNGMuanBlZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI0VDIxMzgwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkMmVkYjFkMTNmODQwMTA4OTA4ZTE0NmQyYWY4N2IzZDA1ZGU0ODA1OWRhOTQ3MjdkOWUwNzg2ZjllYjZkZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uQH-6xww0a9Z9W4LPZRGGChdRTI4aDwTtRY_0eBa50I)
> 
> What we could gain from this is simpler / more compact representation, and also communicate that it's not that the commit content is different but it's the base that actually differs.
> 
> WDYT?

@krlvi is this concept introduce ability to only rebase, right?
Because right now we can only merge and rebase PRs, but not branches.

![image](https://github.com/gitbutlerapp/gitbutler/assets/18498712/1726e09a-6aaa-4592-9846-9e0032d2dbfe)

**Comment by krlvi** [comment]
hmm, this is another rebase - it's when you have performed it and your branch is still in progress, while "rebase and merge" is an operation on completion

**Comment by PavelLaptev** [comment]
@kiril got it. So, if you've merged and rebased the branch, it's considered integrated. Therefore, we should also have a "Rebase branch" option, for example, in this dropdown menu.

![image](https://github.com/gitbutlerapp/gitbutler/assets/18498712/c1de972c-d559-413e-8178-9834c830ea5a)

and then allow to select to which branch we want to rebase. e.g.

<img width="567" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/fb83b8b3-cc04-4eaf-bda7-9f666bab03dc">

**Comment by krlvi** [comment]
Let's discuss tomorrow, I sense there is a confusion here. But in brief, the `Merge Pull Request, Rebase and Merge, Squash and merge` operations that a person does on GitHub are shortcuts for the way the pull request ends up on master.

This is quite different, workflow wise, from rebasing while still working on the branch. For example our update button may perform a rebase.

**Comment by PavelLaptev** [comment]
yes, I understand that `Rebase and Merge`, `Squash and merge` are shortcuts.

So, the `update` button is `rebase` but it rebase commits from on base commit to another right? It's not like on the picture below?

![image](https://github.com/gitbutlerapp/gitbutler/assets/18498712/49b26736-bda7-4c6e-bb3f-17b4b025099e)

**Comment by PavelLaptev** [comment]
I misunderstood @krlvi. We discussed today what the rebase and merge means in terms of virtual branches. @krlvi thanks for visual representations of various corner cases 🖤. I'll share some drafts soon.

<img width="898" alt="image" src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/571fdd7f-341e-4cfe-983d-9ee4c43d56da">

**Comment by PavelLaptev** [comment]
Hey! Made new design drafts here:
https://www.figma.com/file/FbeLt0yjY9RiNn8MXUXsYs/Client-Design?type=design&node-id=2524%3A59821&mode=design&t=BU6UN2A1rkd006La-1

---

_**Here are flows you can find in drafts:**_
- simple local changes push
- merge upstream commits
- rebase upstream commits
- rebase whole workspace

---

[<img src="https://github.com/gitbutlerapp/gitbutler/assets/18498712/bd889a8c-6ef3-45c5-ba35-96b5ae3a41f3">](https://www.figma.com/file/FbeLt0yjY9RiNn8MXUXsYs/Client-Design?type=design&node-id=2524%3A59821&mode=design&t=BU6UN2A1rkd006La-1)

**Comment by krlvi** [comment]
Looking really nice @PavelLaptev! Looks accurate. I will leave some comments in figma on specific things where i had some thoughts. Also, is it worth making some kinda of design that shows the developer that part of the stuff is not on their computer. Not like the dotted line here https://github.com/gitbutlerapp/gitbutler/issues/3121#issuecomment-2007468887 but something 🤔

**Comment by PavelLaptev** [comment]
@krlvi gotcha, thanks! I'll address your comments

**Comment by PavelLaptev** [comment]
Another question is how we want to display unfolded commits.

Currently, we show it like this, in a single lane. Which works, but sometimes it feels a bit squeezed, so you need to resize the lane manually:

https://github.com/gitbutlerapp/gitbutler/assets/18498712/b3646c99-a298-4635-bfd7-12319278f128

---

Another way is to show on the left, like this:

https://github.com/gitbutlerapp/gitbutler/assets/18498712/514af739-4cc1-45b9-82d9-1c10b61c318b

---

What do you think?

**Comment by krlvi** [comment]
hmm, while the second option looks better, i feel that the double expand to the right makes it extra tricky to see the contents at a glance (an additional click and a long travel distance to get to the content of a file).

im worried that it's already a bit tricky to navigate/scroll left/right even for simpler things like moving hunks. 
Are there more options that are possible other than those 2 that don't involve disruptive modals?

**Comment by PavelLaptev** [comment]
Another option could be like this to open is aside, but anyway it will take horizontal space:

https://github.com/gitbutlerapp/gitbutler/assets/18498712/a6305b3e-51a2-40fb-b0ba-722ccc5f95b8

---

And another option where we open the commit card vertically, but show the hunk aside:

https://github.com/gitbutlerapp/gitbutler/assets/18498712/2481ca5b-b553-4dfd-8310-25d870e77535

**Comment by PavelLaptev** [comment]
@krlvi I don't think that moving hunks related to this, because you can move hunks between commits. Now you can only `sqash` the whole commit card

**Comment by PavelLaptev** [comment]
We continue to discuss this on Discrod here https://discord.com/channels/1060193121130000425/1060193121666863156/1222544812998397962

**Comment by PavelLaptev** [comment]
Updated designs are here: https://www.figma.com/file/FbeLt0yjY9RiNn8MXUXsYs/Client-Design?type=design&node-id=2907%3A112917&mode=design&t=HJFGJ2i9aA8qWm2t-1

**Comment by krlvi** [comment]
This is now implemented with v0.12.0 https://discord.com/channels/1060193121130000425/1183737922785116161/1246475064006807575
